[로또] 최재희 미션 제출합니다.

0jenn0 opened the pull request:
![header](https://capsule-render.vercel.app/api?type=rounded&height=120&text=우테코%20프리코스%203주차%20미션%20|%20로또%20🎟️&color=fffded&fontColor=343434&fontSize=35)
<br/>

## 📒 목차

- [📌  로또 게임 설명](#로또-게임-설명)
- [✅  기능 구현 목록](#기능-구현-목록)
  - [1)  로또 구매하기](#1로또-구매하기)
  - [2)  당첨 번호 입력](#2당첨-번호-입력)
  - [3)  결과 발표](#3결과-발표)
- [🚨  로또 번호 재입력 요구 구현 사항](#로또-번호-재입력-요구-구현-사항)
- [🗄️  패키지 구조](#️패키지-구조)
- [🔢  전역 상수 관리하기 (매직 넘버 없애기)](#전역-상수-관리하기-매직-넘버-없애기)

<br/>

## 📌&nbsp;&nbsp;로또 게임 설명

 <table>
<tr>
<th>출력 예시</th>
<th>게임 설명</th>
<tr>
<tr>
<td><img width="300" alt="output" src="https://github.com/woowacourse-precourse/javascript-lotto-6/assets/130737187/2e9cc002-c833-4c2f-afd8-656088ed063a"></td>
<td>로또 구입 후 당첨 통계를 확인하는 게임</td>
</tr>
 </table>


  </br>
  </br>

## ✅&nbsp;&nbsp;기능 구현 목록

### 1)&nbsp;&nbsp;로또 구매하기

- [x] 사용자에게 구입 금액을 입력받는다.

  - [x] 1000원 단위로 입력 받는다.
  - [x] 0, -1000 예외 처리한다.
  - [x] 1,000처럼 천단위 포맷팅 숫자도 입력 받을 수 있게한다.

- [x] 입력받은 금액으로 로또를 생성한다.
  - [x] 'n개를 구매하셨습니다.'라는 문구를 출력한다.
  - [x] 랜덤 로또 번호 생성은`Random.pickUniqueNumbersInRange` 을 이용한다.
- [x] 구매한 로또 번호를 출력한다.
  - [x] 생성된 로또번호는 리스트 형식으로 출력한다.

### 2)&nbsp;&nbsp;당첨 번호 입력

- [x] `Console.readLineSync`를 이용하여 "당첨 번호를 입력해주세요"라는 문구와 함께 입력받는다.
- [x] 사용자에게 당첨 번호를 입력받는다.
  - [x] 로또 숫자는 1~45 사이어야한다.
  - [x] 로또당 번호 갯수는 6개이고 중복되지 않아야한다.
  - [x] ,로 구분하여 입력받는다.
  - [x] 예외 처리 발생시 오류 메시지 출력 후, 재입력을 요구한다.
- [x] 보너스 번호를 입력받는다.
  - [x] 1개만 입력 받아야한다.
  - [x] 당첨 번호와 중복되지 않는지 검증한다.
  - [x] 예외 처리 발생시 오류 메시지 출력 후, 재입력을 요구한다.

### 3)&nbsp;&nbsp;결과 발표

- [x] 당첨 통계를 계산한다.
- [x] 당첨 결과를 출력한다.
- [x] 수익률을 계산한다.
  - [x] 소숫점 둘째자리에서 반올림한다.
- [x] 수익률을 출력한다.
  - [x] 천 단위로 , 포맷팅을 한다.
  - [x] 소숫점 첫째짜리까지 출력한다.

<br/>
<br/>

## 🚨&nbsp;&nbsp;로또 번호 재입력 요구 구현 사항

<table style="font-size:16px">
<tr>
  <th>당첨번호 번호 예외 발생 시</th>
  <th>보너스 번호 예외 발생 시 </th>
</tr>
<tr>
  <td>
<img width="281" alt="lottoNumbersRetry" src="https://github.com/woowacourse-precourse/javascript-lotto-6/assets/130737187/b8a71bbc-d763-47a5-9810-94e64311884a">
  </td>
  <td>
<img width="337" alt="bonusNumberRetry" src="https://github.com/woowacourse-precourse/javascript-lotto-6/assets/130737187/d3ff96cc-b264-4210-a638-e17fd3eb448b">
  </td>
</tr>
<tr>
<td>로또 번호에 대한 예외 발생이 안될 때 까지 재입력을 요구하도록 구현</td>
<td>다시 <b>당첨 번호 입력으로 돌아가지 않고</b> 보너스 번호 재입력을 요구하도록 구현</td>
</tr>
</table>



 <br/>
<br/>

## 🗄️&nbsp;&nbsp;패키지 구조

<table style="width:90%">
  <tr>
    <th >Package</th>
    <th align="center">Class</th>
    <th align="center">Description</th>
  </tr>
  <tr>
  <td><b>config</b></td>
  <td><b>LottoSettings</b></td>
  <td>로또 게임 설정을 관리</td>
  <tr>
  <tr>
    <td rowspan="4"><b>controller</b></td>
    <td><b>LottoGame</b></td>
    <td>게임의 메인 흐름을 제어</td>
  </tr>
  <tr>
    <td><b>LottoPurchaseController</b></td>
    <td>로또 구매를 담당</td>
  </tr>
  <tr>
    <td><b>DrawnLottoController</b></td>
    <td>추첨된 로또 생성을 담당</td>
  </tr>
  <tr>
    <td><b>AnnounceResult</b></td>
    <td>결과 발표를 담당</td>
  </tr>
  <tr>
    <td rowspan="3"><b>domain</b></td>
    <td><b>Lotto</b></td>
    <td>로또 번호를 관리</td>
  </tr>
  <tr>
    <td><b>PurchaseAmount</b></td>
    <td>구매 금액을 관리</td>
  </tr>
  <tr>
    <td><b>DrawnLotto</b></td>
    <td>추첨된 로또 번호를 관리</td>
  </tr>
<tr>
<td><b>service</b></td>
<td><b>LottoYieldCalculator</b></td>
<td>당첨 통계과 수익률을 계산</td>
</tr>
  <tr>
    <td rowspan="2"><b>view</b></td>
    <td><b>InputView</b></td>
    <td>사용자 입력을 처리</td>
  </tr>
  <tr>
    <td><b>OutputView</b></td>
    <td>출력을 담당</td>
  </tr>
<tr>
<td rowspan="3"><b>validator</b></td>
<td><b>LottoValidator</b></td>
<td>로또 번호를 검증</td>
</tr>
<tr>
<td><b>PurchaseAmountValidator</b></td>
<td>구입 금액을 검증</td>
</tr>
<tr>
<td><b>DrawnLottoValidator</b></td>
<td>당첨 번호와 보너스 번호를 검증</td>
</tr>


<tr>
<th colspan="3" style="text-align: center;" >flow chart</th>
</tr>

<tr>
<td colspan="3">
<img src="https://github.com/woowacourse-precourse/javascript-lotto-6/assets/130737187/c201bc41-7624-4ceb-a0cb-49ae35901f02"/>
</td>
</tr>


</table>

</br>
</br>

## 🔢&nbsp;&nbsp;전역 상수 관리하기 (매직 넘버 없애기)

> <b>LottoSettings</b> 클래스는 <b>LOTTO_SETTINGS</b>(로또 게임의 설정과 관련된 상수 오브젝트)를 캡슐화

<table style="width:90%">
  <tr>
    <th align="center">Package</th>
    <th align="center">Code</th>
    <th align="center">Description</th>
  </tr>
  <tr>
    <td><b>constants</b></td>
    <td>
<img src="https://github.com/woowacourse-precourse/javascript-lotto-6/assets/130737187/4d3aa684-697f-43d2-879f-053107898f4c">
</td>
  <td>deepFreeze 함수는 설정 객체의 깊은 불변성을 보장하여, 게임 실행 중 의도치 않은 변경을 방지</td> 
    
  </tr>
  <tr>
  
  </tr>

  <tr>
    <td rowspan="4"><b>Class</b></td>
    <td rowspan="4"><img src="https://github.com/woowacourse-precourse/javascript-lotto-6/assets/130737187/827e96e4-41e1-4e60-89fc-4dd833df2023"></td> 
      <th>중앙 집중식 설정 관리 </th>
  </tr>
   <tr>
   <td>클래스를 통해 모든 설정을 관리하여 설정 값의 중복 사용을 방지하고 일관성을 유지</td>
  </tr>
  </tr>
  <tr>
  <th>유지보수 용이성 </th>
  </tr>
   <tr>
   <td> 설정 값이나 게임 규칙 변경 시 LottoSettings 클래스만 수정하면 된다</td>
  </tr>

</table>



</br>
</br>

0jenn0 committed:
Docs: README에기능 구현 목록 작성

0jenn0 committed:
Feat: 커스텀 에러 추가

커스텀에러로 에러 메시지 앞에 [ERROR]라는 prefix 붙게함

0jenn0 committed:
Feat: 로또게임 세팅(상수들) 추가

- 게임에서 사용되는 의미있는 숫자들을 모두 gameSetting.js에 모아 관리할수있도록함.
- 객체 안의 객체도 얼리는 deepFreeze 함수 추가함.

0jenn0 committed:
Refactor: 로또 번호가 6개를 초과할 경우 에러 처리 부분 커스텀 에러로 수정 및 상수 이용

커스텀에러 LottoError를 사용하도록 수정하고 6개라고 하드코딩되어있던 부분도 상수를 이용함

0jenn0 committed:
Refactor: Lotto의 숫자 개수 검증 함ㅅ무에서 커스텀 에러과 상수 사용

커스텀 에러를 사용하고 6개 하드코딩된 부분을 상수를 이용하도록 리팩토링함

0jenn0 committed:
Refactor: Lotto.js를 src -> src/Models 로 이동시킴

0jenn0 committed:
Feat: Lotto의 숫자들 오름차순 기능 추가

번호를 받아 오름차순으로 정리한 후 Lotto 인스턴스가 생성되도록 함

0jenn0 committed:
Feat: Lotto 중복된 숫자 검증 추가

0jenn0 committed:
Test: Lotto 메서드들 기능 테스트 추가

- 로또번호가 오름차순으로 정렬되어 나오는지 확인하는 테스트코드 추가

0jenn0 committed:
Fix: LottoError에서 잘못된 클래스 참조 수정

0jenn0 committed:
Feat: Lotto의 number값을 받아오는 함수 추가

Lotto의 numbers는 private이기 때문에 값을 얻기위해서는 따로 메서드 추가가 필요함

0jenn0 committed:
Fix: 로또 번호 오름차순 테스트 코드 안의 잘못된 expect 구문 수정

0jenn0 committed:
Feat: Lotto 번호 범위 검증 함수 추가

0jenn0 committed:
Test: Lotto 번호 범위 검증 메서드 테스트 추가

0jenn0 committed:
Docs: 기능 구현 목록에 Lotto 숫자 범위 검증하기 추가 및 완료 체크

번호가 1이상 45이하인지 검증하는거 추가하고 구현 후 테스트 통과도했기에 완료 체크함

0jenn0 committed:
Feat: Money에 입력받은 숫자에서 ,나 .가 있으면 제거하기 추가

1,000 1.000 이런 포맷의 금액도 받을 수 있도록 추가함

0jenn0 committed:
Feat: Money의 #money값 받아오는 함수 getMoney 추가

0jenn0 committed:
Test: Money 클래스의 포맷팅되어있는 숫자 언포맷팅하는 메서드 테스트 추가

1,000이나 1.000도 1000으로 받을 수 있게하는 함수를 테스트하는 코드 추가

0jenn0 committed:
Feat: Money 클래스에 전달 받은 값이 숫자 아니면 에러 처리하기 추가

,와 .을 제거한 후에도 isNaN이면 에러처리하게함

0jenn0 committed:
Money 클래스에서 숫자가 아닌값 입력받으면 에러처리하기 테스트 코드 추가

0jenn0 committed:
Money 클래스에 입력받은 금액이 1000단위가 아니면 에러 처리 추가

0jenn0 committed:
Test: Money가 1000 단위가 아닌 숫자값을 입력받으면 에러처리 테스트 코드 추가

0jenn0 committed:
Fix: Money에서 1000단위 검증 메서드에 단,금액이 0보다 커야된다는 조건 추가

0jenn0 committed:
Test: Money 클래스의 1000단위 검증 테스트 케이스 추가

1000으로 나누어 떨어져도 0과 -1000은 안되므로 이 케이스들에 대한 테스트도 추가함

0jenn0 committed:
Refactor: 하드코딩해놨던 6을 상수 사용함

0jenn0 committed:
Docs: 기능 구현 목록에 Lotto 로또 번호 범위 검증 추가, DrawnLotto 클래스 추가

당첨번호 6개와 보너스 번호 1개를 받는 당첨로또를 Lotto를 상속받은 DrawnLotto로 구현하기로함

0jenn0 committed:
Feat: DrawnLotto를 만들고 그 안에 당첨번호와 보너스가 중복되는지 검증하는 메서드 추가

Lotto를 상속받아 당첨번호는 Lotto의 검증을 거치고 그 후 보너스번호(bonusNumber)가 당첨번호(numbers)와 중복되는지 검증하여

0jenn0 committed:
Feat: 당첨번호와 보너스 번호가 필드인 DrawnLotto 추가

6개번호 + 보너스 번호로 이루어져있는 당첨로또(DrawnLotto)를 Lotto를 상속받아 구현

0jenn0 committed:
Feat: DrawnLotto에보너스숫자와 당첨번호가 중복되는지 검증하는 메서드 추가

0jenn0 committed:
Test: Drawn 클래스 테스트에 당첨번호와 보너스 번호의 중복 검증 메서드 테스트 코드 추가

0jenn0 committed:
Feat: DrawnLotto 클래스에 보너스 번호의 범위 검증 메서드 추가

0jenn0 committed:
Test: DrwanLotto 클래스에 보너스 번호 범위 검증 메서드 테스트 코드 추가

0jenn0 committed:
Docs: 기능 구현 목록의 DrawnLotto에 항목 추가 및 완료체크

- 보너스 번호가 1이상 45이하인지 검증하기 목록에 추가
- DrwanLotto의 두 기능에 대한 구현 후 테스트 코드 통과하여 완료체크함

0jenn0 committed:
Feat: LottoYieldCalculator에 로또번호,당첨번호 일치 갯수 계산하기 추가

0jenn0 committed:
Feat: DrawnLotto에 numbers와 bonusNumber를 객체로 리턴하는 메서드 추가

0jenn0 committed:
Fix: getFullNumbers에서 변수 선언할 때 const 빼먹은것 수정

0jenn0 committed:
Test: DrawnLotto에서 numbers와 bonusNumber를 객체로 리턴 메서드 테스트 코드 추가

0jenn0 committed:
Fix(test) : DrawnLotto 테스트 코드들 number인풋들 string인풋으로 수정

0jenn0 committed:
Lotto에 입력받은 숫자들을 number로 형변환 추가

0jenn0 committed:
Fix: DrawnLotto 클래스,Lotto 클래스 number로 형변환 추가

input에서 string으로 가져올거기 때문에 이 클래스들에서 number로 형변환 해줘야한다

0jenn0 committed:
Fix(test): LottoTest 테스트 케이스 input들 string으로 수정

0jenn0 committed:
Test: LottoYieldCalculator 클래스의 번호 5개 맞으면 보너스 번호초 체크하는지 테스트 코드 추가

0jenn0 committed:
Refactor: 로또 번호 일치 갯수 구하는 메서드 분리

0jenn0 committed:
Refactor: LOTTO_SETTINGS의 PRIZE_FOR_TIER -> PRIZES로 수정

0jenn0 committed:
Feat: 로또번호와 당첨번호를 받아 일치갯수 구하는 메서드 추가

0jenn0 committed:
Feat: 일치갯수를 인자로 받아 상금을 구하는 메서드 추가

0jenn0 committed:
Test: LottoYeildCaculator에서 로또 1등하는 경우 테스트 코드 추가

0jenn0 committed:
Fix: LottoYieldCalculator에서 참조값 수정

LOTTO_SETTINGS에서 NUMBERS_PER_TICKET, PRIZES를 가져오기위한 참조값 잘못 적었던것 수정함

0jenn0 committed:
Test: LottoYeildCaculator 클래스 테스트에 1등~5등,당첨금 못받는 케이스 테스트 코츠 추가

0jenn0 committed:
Fix: LottoYieldCalculator의 2등 상금 구하는 부분 수정

- 기존의 코드는 3등인 부분도 matchCount === 5 인 부분에서 걸려서 return 당하기때문에 2등 상금은 잘 구해지지만 3등인 경우는 undefined로 나오는 오류 생김
- 일치갯수가 5이고 보너스 번호도 맞는 경우라고 정확하게 수정하여 오류 해결함

0jenn0 committed:
Feat: 로또 뭉치들과 당첨번호 비교해서 총 당첨금 구하는 메서드 추가

0jenn0 committed:
Test: LottoYieldCalculator의 로또가 여러 개일 때도 총 당첨금액 구하는 메서드 테스트 코드 추가

0jenn0 committed:
Feat: LottoYieldCalculator에 로또 비용에 대한 총당첨금의 수익률 계산하는 메서드 추가

0jenn0 committed:
Test: LottoYieldCalculator의 로또 비용에 대한 총당첨금 수익률 계산하는 메서드 테스트 코드 추가

0jenn0 committed:
Docs: 기능 구현 목록에 calculaor 목록 완료체크

- 해당 기능을 LottoYieldCalculator.js에 구현 후 테스트 통과했기에 완료체크함

0jenn0 committed:
Feat: InputView에 사용자에게 로또 구입 금액 입력받기 추가

0jenn0 committed:
Feat: InputView에 로또 당첨 번호 입력받기 추가

0jenn0 committed:
Feat: InputView에 로또 보너스 번호 입력받기 추가

0jenn0 committed:
Test: InputView의 로또 당첨번호 입력받기 테스트 코드 추가

0jenn0 committed:
Test: InputView 로또 보너스 번호 받기 추가

0jenn0 committed:
Test: InputView의 로또 구입 금액 입력 받기 테스트 코드 추가

0jenn0 committed:
Fix: InputView 메서드들 오타 수정 및 처리 기능 추가

- promptToBuyLottos 메서드 속 trim() 오타 수정
- promptForDrawnLottoNumbers에서 맨 뒤에 (,) 있을시 제거 공백 제거 추가
- promptForBonusNumber에도 공백 제거 추가

0jenn0 committed:
Feat: InputView에서 쓰인 메시지들 상수화

0jenn0 committed:
Docs: 구현 기능 목록의 inputView 목록에 보너스 번호 입력 추가 및 완료체크

- inputView 목록에 보너스 번호 입력 추가 목록 추가
- inputView의 3가지 목록을 다 구현 후 테스트 통과했기에 완료체크함 ✔️

0jenn0 committed:
Feat: LOTTO_SETTINGS에mathchNum도 추가

- 각 등수마다 일치갯수 NUMBERS_PER_TICKET로 나타낼려다가 가독성이 떨어지고 이렇게하는게 로또 당첨갯수 구하는 함수 구현하는데 더 편할것같아서 수정함

0jenn0 committed:
Refactor: 일치갯수 계산하는 메서드에서 보너스 번호 가지고있는지도 함께 객체로 리턴하도록 수정함

0jenn0 committed:
Refactor: 구매 로또들 등수 결과 구하는 부분 리팩토링함

- result 배열을 하드코딩하지 않고 Object메서드를 이용하여 생성

0jenn0 committed:
Feat: LottoYieldCalculator 수익률 계산하기 추가

- getResult : 로또 1개가 아니라 로또 여러개 받아서 결과 리턴하게 수정
- getResult로 등수별 갯수 받아서 최종 상금액에 대한 수익률 구하기 추가

0jenn0 committed:
Feat: 로또 등수 이름 상수화 추가

- 코드 내에 로또 등수 이름자체를 써야할 때가 있어서 이 때, 하드코딩하지 않고 오타로 인한 오류 발생 방지하기위해 이름도 상수화하여 타입처럼 강제함

0jenn0 committed:
Feat: 당첨 통계 출력하기 추가, 구매한 로또 번호들 출력하기 추가

0jenn0 committed:
Fix: 경로뒤에 .js 빠진거 수정함

0jenn0 committed:
Refactor: 티켓 가격 1000원 하드코딩되어있던것 수정

0jenn0 committed:
Fix: 경로 뒤 .js 빠진것 추가

0jenn0 committed:
Refactor: DrawnLotto 생성자에서 bonusNumber 밖으로 뺌

사용자에게 당첨번호부터 입력받고 보너스 번호를 입력받는데 이 때 당첨번호부터 검증하고 그 뒤에 보너스 번호받으면
그 때가서 또 검증해서 그 때 그 때 재입력 받을 수 있도록 bonumnNumer을 set하는 부분과 검증부 constructor에서 빼냄

0jenn0 committed:
Style: 메시지 스타일 수정

0jenn0 committed:
Feat: LottoGame에 로또 구매할 금액 입력받기 추가

예외처리가 나면 재입력 받도록 구현함

0jenn0 committed:
Feat: LottoGame에 구매한 금액대로 로또 생성 추가

0jenn0 committed:
Feat: LottoGame에 당첨번호 입력받기 추가

예외처리가 될 경우 재입력 받도록 구현함

0jenn0 committed:
Feat: LottoGame에 보너스 번호 입력받기 추가

예외처리 당할 경우 재입력 받도록 구현함

0jenn0 committed:
Feat: LottoGame에 로또 당첨 통계 구하고 출력하기 추가

0jenn0 committed:
Feat: App play안에 LottoGame start() 추가

0jenn0 committed:
Refactor: LottoYieldCalculator에 setting DI함

0jenn0 committed:
Refactor: Money -> PurchaseAmount로 수정, getNumberOfLottos 추가

- 클래스 이름과 메서드 이름을 수정하고 로또구매 갯수도 리턴하는 함수 추가

0jenn0 committed:
Refactor: LottoYieldCalculator 메서드들 분리

- 길어진 메서드들 기능별로 분리함
- calculatePrizeForRank : rank의 prize에 matchcount곱해서 prize 계산하는 부분 분리함
- initialiizeResultObject : result 오브젝트를 setting을 이용하여 만드는 부분 분리함

0jenn0 committed:
Refactor: Lotto클래스와 DrawnLotto 클래스에서 검증 부분 Validator클래스로 분리

0jenn0 committed:
Refacotr: Validator들에게 setting 의존성 주입함

0jenn0 committed:
Refactor: LOTTO_SETTINGS 상수의 추상화 (Models,Validator만)

- 여러 클래스들에서 LOTTO_SETTINGS 상수를 가져다쓰면서 여러 메서드 구현이
LOTTO_SETTINGS의 구조에 의존하여 구현을 했기에 LOTTO_SETTINGS의 상수명이나 구조가 바뀔 시 해당 부분들 수정 불가피하였음.
- 따라서 LottoSettings라는 클래스를 만들어 LOTTO_SETTINGS를 추상화하였음.
- Views는 View자체도 추상화하고 그 뒤에 LOTTO_SETTINGS 추상화도 추가해줄 예정

0jenn0 committed:
Refactor: LottoGame 메서드 분리

0jenn0 committed:
Refactor: LottoGame을 여러개의 controller로 분리함

-LottoPurchaseController: 로또 금액 입력부터 구입한 로또번호 출력까지 컨트롤
-DrawnLottoController: 당첨 로또 번호 입력부터 보너스 입력까지 컨트롤
-AnnounceResultController: 당첨 결과과 수익률을 계산하고 각 결과 출력까지  컨트롤
-LottoGame: 위 3개의 controller간의 값 이동과 실행 순서를 컨트롤

0jenn0 committed:
Refactor: prompt에서 에러 메시지 출력에 쓰인   Console.print()를 OutputView의 print메서드로 추상화

0jenn0 committed:
Refactor: AnnounceResult 당첨 결과,수익률을 같이 print하도록 수정

등수별 당첨 결과와 수익률 사이에 빈 줄 없게하기 위해 수정함

0jenn0 committed:
Refactor: LottoPurchaseController에서 랜덤으로 번호 만들기 부분 함수로 분리

- 분리하면서 1,45,6 하드코딩되어있던 숫자들도 상수사용 함

0jenn0 committed:
Refactor: LottoYieldCalculator 함수 인자수 줄이기, for of 안쓰게 수정

- 함수 인자에 drawnLottoNumbers, bonusNumber 이렇게 풀어서 넣게하지않고, drawnLotto를 받게 한 후 코드내에서 사용할 값 받아오게 수정
- for .. of를  사용하지 않도록 수정함

0jenn0 committed:
Docs: DrawnLotto 예외 경우 추가

- 보너스 번호 2개 이상일 경우 예외처리 추가함

0jenn0 committed:
Feat: DrawnLottoValidator에 보너스 번호 isNaN 검증 추가

보너스 번호를 2개이상 입력 받을 경우 중간에 스페이스나 ,같은 기호가 있어 isNaN에 걸릴것이므로 isNaN으로 예외처리 추가함

0jenn0 committed:
Refactor: 수익률 천단위 , 포맷팅 추가

천단위로 ,로 포맷팅하고 소숫점 첫째짜리 까지 보여주도록 포맷팅 추가함

0jenn0 committed:
Refactor: 상수 하드코딩 LottoSetting 사용으로 수정, 긴 메서드 분리

- printResultMessage에서 당첨결과 메시지 생성하던 부분을 createWinningResultMessage라는 메서드로 분리시킴

0jenn0 committed:
Test: DrawnLotto 클래스 기능 테스트 코드 추가함

- 예외처리가 잘되는지 테스트하는 코드 추가함

0jenn0 committed:
Fix: 보너스 번호 검증 후 number로 형변환 추가₩

0jenn0 committed:
Test: Lotto 클래스 기능 테스트 코드 추가

- 로또 번호 검증과 오름차순 정렬,엣지케이스 테스트 코드 추가함

0jenn0 committed:
Test: Money->PurchaseAmount 클래스 이름,메서드 이름 변경 사항 수정

0jenn0 committed:
Docs: 기능 구현 목록 수정 및 항목 추가

- 객체지향적이던 기능 구현 목록을 절차지향적으로 수정
- 로또 게임 설명,로또 번호 재입력 요구 구현사항, 패키지 구조, 매직 넘버 없애기 항목 추가

0jenn0 committed:
Docs: README 비어져있던 내용 추가

0jenn0 committed:
Refactor: Lotto.jsfmf  src/로 이동

0jenn0 committed:
Test: Lotto 위치 변경에 따른 수정

brgndyy commented on src/Lotto.js:
```diff
@@ -1,18 +1,24 @@
   constructor(numbers) {
-    this.#validate(numbers);
-    this.#numbers = numbers;
+    const convertedNumbers = numbers.map(Number);
```
아예 여기서 sort 를 해주어도 괜찮지 않을까요? 

저는 validator 클래스를 따로 사용하시는김에 type 검사 기능을 추가해주신다면 

this.#sortNumbers()라는 필드 없이도 구현이 가능할거라고 생각하는데, 따로 this.#sortNumbers() 를 선언해주신 이유가 있으실까요 ? 의견 여쭤보고 싶습니다!

0jenn0 replied:
> 아예 여기서 sort 를 해주어도 괜찮지 않을까요?
> 
> 저는 validator 클래스를 따로 사용하시는김에 type 검사 기능을 추가해주신다면
> 
> this.#sortNumbers()라는 필드 없이도 구현이 가능할거라고 생각하는데, 따로 this.#sortNumbers() 를 선언해주신 이유가 있으실까요 ? 의견 여쭤보고 싶습니다!

validator 클래스는 말그대로 검증하는 클래스이고, 오름차순 정렬은 로또 번호가 가지는 속성이기때문에 로또 클래스에서 처리해줘야겠다고 생각했습니다 !

brgndyy commented on src/Controllers/DrawnLottoController.js:
```diff
@@ -0,0 +1,42 @@
+    let drawnLottoWithoutBonusNumber;
+    while (true) {
+      try {
+        const inputdrawnLottoNumbers =
```
while- try-catch 가 반복되는것 같은데, 이 틀 자체를 따로 유틸함수로 만들어주어도 괜찮을것 같다고 생각이 들어요!

brgndyy commented on src/Validator/DrawnLottoValidator.js:
```diff
@@ -0,0 +1,39 @@
+
+  #validateBonusNumberDuplicate(numbers, bonusNumber) {
+    if (numbers.includes(bonusNumber))
+      throw new LottoError("보너스 번호가 당첨번호와 중복됩니다.");
```
메세지들을 따로 상수처리 해주시는게 어떨지 의견 여쭙습니다.

2주차 공통 피드백중에서도 "값을 하드 코딩하지 않는다" 라는 말이 있었어서요 :)

0jenn0 replied:
로또 관련 상수 처리하다가 저런 메시지 부분에서 꼼꼼히 못한것 같습니다 😓
메세지들도 상수처리 하는게 좋았을거같습니다 !

brgndyy commented on src/Error/LottoError.js:
```diff
@@ -0,0 +1,8 @@
+export default class LottoError extends Error {
```
LottoError 라고 명명 되어있으니 처음 봤을때, 로또 번호에만 관련한 커스텀 에러 클래스라는 생각이 들었어요! 

총체적인 AppError나 다른 명명이 더 직관적으로 와닿지 않을까? 하는 생각이 듭니다! 🫡 (철저히 개인적인 의견이니 듣고 넘겨주셔도 괜찮습니다.)

0jenn0 committed:
Test: LottoSettings 클래스 기능 테스트 코드 추가함

- 각 메서드들이 예상대로 리턴을 잘하는지 테스트 하는 코드 추가함
-

0jenn0 committed:
Test: LottoSettings 클래스 기능 테스트 코드 추가

- 각각의 메서드가 LOTTO_SETTING에서 값을 찾아 알맞게 리턴하는지 테스트하는 코드 추가
- LOTTO_SETTINGS의 객체 안의 객체까지 깊게 얼려져있는지 테스트하는 코드 추가

brgndyy commented on src/config/gameSetting.js:
```diff
@@ -0,0 +1,72 @@
+});
+
+export class LottoSettings {
+  settings;
```
혹시 로또셋팅 안에 들어가는 값들이 상수 값들인것 같은데, 이런식으로 클래스화 해서 나타내주신 이유가 있으실까요 ?

한수 배우고 싶은 마음에 궁금해서 여쭈어 봅니다 ! 🫡

0jenn0 replied:
> 혹시 로또셋팅 안에 들어가는 값들이 상수 값들인것 같은데, 이런식으로 클래스화 해서 나타내주신 이유가 있으실까요 ?
> 
> 한수 배우고 싶은 마음에 궁금해서 여쭈어 봅니다 ! 🫡

저도 처음에는 로또 상수 객체만 두고 시작했는데,
구현하는 과정에서

```javascript
const firstPrize = LOTTO_SETTINGS.WINNINGS.FIFTH_PRIZE

```
이렇게 그냥 상수를 가져다쓰면 상수 객체의 구조가 바뀌거나 상수명이 바뀔경우 이 부분이 쓰인 모든 곳에서
다 수정을 해줘야하기때문에 그럴 일이 안생기도록 추상화 하는 클래스를 만들게되었습니다

brgndyy commented on src/config/gameSetting.js:
```diff
@@ -0,0 +1,72 @@
+  }
+}
+
+function deepFreeze(object) {
```
저같은 경우는 deepFreeze 함수 자체를 따로 유틸 함수 폴더로 빼서 관리해주었는데요.

게임 셋팅보다 더 넓은 의미, 용도를 갖고 있는 함수라 생각하는데 다른 폴더로 관리해주시는건 어떻게 생각하실까요 ?

0jenn0 replied:
> 저같은 경우는 deepFreeze 함수 자체를 따로 유틸 함수 폴더로 빼서 관리해주었는데요.
> 
> 게임 셋팅보다 더 넓은 의미, 용도를 갖고 있는 함수라 생각하는데 다른 폴더로 관리해주시는건 어떻게 생각하실까요 ?

이 폴더에 있는것보다 유틸 폴더에 들어가있는게 더 깔끔한것같습니다👍

0jenn0 committed:
Test: LottoYieldCalculator 클래스 테스트 코드 추가

- 총 상금계산,수익률 계산,당첨된 로또 결과 리턴 기능 테스트 코드 추가

0jenn0 committed:
Fix: findPrizeKey 메서드에서 setting 빼먹은것 수정

0jenn0 committed:
Test: Models 예외처리 제외 각 클래스의 get 메서드 기능 테스트 코드 추가

- PurchaseAmount,Lotto,DrawnLotto에서 자기 필드 값을 리턴하는 메서드들에 대한 테스트 코드 추가함

0jenn0 committed:
Refactor: LottoTest 파일 위치 원래 위치로 이동

brgndyy reviewed (commented):
안녕하세요 재희님 pr 구경하다가 우연히 들리게 되었습니다.

리뷰 원치 않으셨을수도 있지만 약소하게 남기고 갑니다 :) 

괜찮으시다면 제 pr도 한번 보러 와주시면 감사하겠습니다. 

4주차 마무리도 잘 부탁드립니다 

https://github.com/woowacourse-precourse/javascript-lotto-6/pull/30
